text/0000-gen-fn.md

```diff
@@ -0,0 +1,429 @@
+
+Like other uses of `impl Trait`, auto traits are revealed without being specified.
+
+### `#[rustc_gen]` blocks
+
+`#[rustc_gen]` blocks are the same as an unstable generator
+
+* without arguments,
+* with an additional check forbidding holding borrows across `yield` points,
+* and an automatic `Iterator` implementation.
+
+We'll probably be able to modularize the generator implementation and make it more robust on the implementation and diagnostics side for the `#[rustc_gen]` block case, but I believe the initial implementation should be a HIR lowering to a generator and wrapping that generator in [`from_generator`][].
+
+## Fusing
+
+Special case the generator implementation if `Generator::Return = ()` to not panic, but
+repeatedly produce values of the unit type.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+It's another language feature for something that can already be written entirely in user code.
+
+In contrast to `Generator`, `#[rustc_gen]` blocks that produce `Iterator`s cannot hold references across `yield` points.
+See [`from_generator`][] which has an `Unpin` bound on the generator it takes to produce an `Iterator`.
+
+[`from_generator`]: https://doc.rust-lang.org/std/iter/fn.from_generator.html
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+## Keyword
+
+We could use `iter` as the keyword.
+I prefer `iter` because I connect generators with a more powerful scheme than plain `Iterator`s.
+The `Generator` trait can do everything that `iter` blocks and `async` blocks can do and more.
+I believe connecting the `Iterator` trait with `iter` blocks is the right choice,
+but that would require us to carve out many exceptions for this keyword as `iter` is used for module names and method names everywhere (including libstd/libcore).
+It may not be much worse than `gen` (see also [the unresolved questions][#unresolved-questions]).
+We may want to use `gen` for full on generators in the future.
+
+## Do not do this
+
+One alternative is to keep adding more helper methods to `Iterator`.
+It is already hard for new Rustaceans to be aware of all the capabilities of `Iterator`.
+Some of these new methods would need to be very generic.
+While it's not an `Iterator` example, [`array::try_map`][] is something that has very complex diagnostics that are hard to improve, even if it's nice once it works.
+
+Users can use crates like [`genawaiter`](https://crates.io/crates/genawaiter) or [`propane`](https://crates.io/crates/propane) instead.
+`genawaiter` works on stable and provides `gen!` macro blocks that behave like `#[rustc_gen]` blocks, but don't have compiler support for nice diagnostics or language support for the `?` operator. The `propane` crate uses the `Generator` trait from nightly and works mostly
+like `#[rustc_gen]` would.
+
+[`array::try_map`]: https://doc.rust-lang.org/std/primitive.array.html#method.try_map
+
+## `return` statements `yield` one last element
```

Comment:
I find it funny that none of the concrete examples in this subsection actually use explicit `return` statements.

Is there a potential world where we would continue to treat a trailing `()` expression as terminating the iteration with no associated yield, while an *explicit* `return expr` would yield the value of `expr` and then terminate the expression? Would that world address any/all of the concerns raised below about trailing expressions? (Would that "benefit" justify the damage wrought by now introducing a distinction between trailing expressions vs `return` that we had not previously had?)

Reply:
I don't think that distinction would help, as then you'd rule out early abort via `return`, unless you special case `return ()` which gets really weird if your iterator also yields `()`

Reply:
If you have an explicit return point where you want to yield one final element that can be written `return yield foo;` (given my assumption of how the desugaring will look).

Reply:
> as then you'd rule out early abort via return

oh, maybe I wasn't clear. By "and then terminate the expression", I *meant* to still early abort. Yield the returned value and don't iterate furhter.

Reply:
We've made a note of this in the relevant section of the RFC.